### PR TITLE
4 setup database interactions class

### LINKED ISSUE #4: Setup Database Interactions class
- [x] Create Database helper class that deals with:
  - connecting/disconnecting to the database (via dsn string or otherwise)
  - updating rows
  - creating rows
  - deleting rows
    - deleting a user should mark it as deleted **NOT** remove it from the relation.
- [x] Initialize User, Product, Order tables
  - User:
    - id: INT UNIQUE NOT NULL PRIMARY KEY
    - username: VARCHAR NOT NULL
    - email: VARCHAR NOT NULL
    - password: VARCHAR NOT NULL
    - deleted: BOOLEAN NOT NULL
  - Product: 
    - id: INT UNIQUE NOT NULL PRIMARY KEY
    - name: VARCHAR NOT NULL
    - description: VARCHAR NOT NULL
    - price: REAL NOT NULL - Find a more "efficient" data type (i.e. one with only 2 decimal places accuracy or something)
    - quantity: INTEGER NOT NULL
    - **ensure quantity >= 0 and price >= 0.0**
  - Order:
    - user_id: INT UNIQUE NOT NULL
    - prod_id: INT UNIQUE NOT NULL
    - quantity: INT NOT NULL
    - PRIMARY KEY (user_id, prod_id)
    - FOREIGN KEY user_id, prod_id REFERENCES User(id), Product(id)
    - **ensure quantity >= 0**
  - [ ] Query Order to get all product, quantity pairs that user (which may have been deleted) has purchased.
- [ ] Add javadocs were needed.

### PR DESCRIPTION
**Pull request by userBD10**
- [x] Create Database helper class that deals with:
   * connecting/disconnecting to the database (via dsn string or otherwise)
   * updating rows
   * creating rows
   * deleting rows
      * deleting a user should mark it as deleted NOT remove it from the relation.
- [x] Initialized and Integrated the User and Product Services with PostgreSQL instead of writing/reading from files
   * Includes the CRUD operations - create, update, delete, get
- [x] We're only connecting to one database, such that both users and products tables are in one database.

## What else needs to be done
- [x] The Order Service (for Yash and Navroop)
- [ ] Code clean-up (I'll do this specifically organizing the helper classes for the database)
- [ ] Make sure the ISCS is working properly with the services (for Mars)
- [ ] Make sure the attributes of each table adhere the data types specified for optimization (I'll do this too)

## How to run
1. Make sure you have a Dockerfile, which is in the database folder that Yash created.
2. If you want to test locally, make sure you have Docker installed. And if you want to run in the lab machines, change the url string from "localhost" to the IP specified in config.json.
3. Follow the README.txt in the csc301_a2 folder.
4. To open the database to make sure that the data is actually being stored, run this command: **docker exec -it assignmentpostgrescontainer2 psql -U assignmentuser -d assignmentdb**
    * the "assignmentpostgrescontainer2" might differ depending on what you name your container when setting up the Dockerfile

## Further Clarifications
Since A2 Component 2 will not have the restart and shutdown functionalities, we can remove these functions from the services. The data in the database will automatically persists so no need to handle that. Correct me if I'm wrong tho or if I'm misunderstanding it.

### CODE REVIEW

**Review by marsponce — CHANGES_REQUESTED** [review]
remove .idea from the commit

File: csc301_a2.iml
Comment (marceloponceardon) [inline csc301_a2.iml:1]: what is this for?

File: runme.sh
Comment (marceloponceardon) [inline runme.sh:1]: instead of commenting out old commands just delete them

File: src/OrderService/OrderDatabase.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+    /**
+     * Retrieves a user's information from the database based on the user ID.
+     * @param id is the ID of the user.
+     * @return A JSON string containing the products ID as a key and quantity as a value
+     */
+    public String getPurchased(int user_id) {
+        String sql = "SELECT prod_id, quantity FROM orders WHERE user_id = ?";
+        try (Connection con = this.connect();
+             PreparedStatement statement = con.prepareStatement(sql)) {
```
Comment (marceloponceardon) [inline src/OrderService/OrderDatabase.java:100]: nice

File: src/OrderService/OrderDatabase.java
Comment (marceloponceardon) [inline src/OrderService/OrderDatabase.java:1]: Make sure you are using TABS not SPACES to indent your code
please or i'll cry
It is definitely an IDE setting you can set, just google it

File: src/OrderService/OrderService.java [hunk 2]
```diff
@@ -117,114 +118,135 @@ public void handle(HttpExchange exchange) throws IOException
...
+                    //     for (int i = 0; i < product_jsonArray.length(); i++)
+                    //     {
+                    //         if (product_jsonArray.getJSONObject(i).getInt("id") == (jsonObject.getInt("product_id")))
+                    //         {
+                    //             //make sure the updated quantity (post order) can't go below zero
+                    //             int updated_quantity = product_jsonArray.getJSONObject(i).getInt("quantity") - jsonObject.getInt("quantity");
+                    //             if (updated_quantity >= 0)
+                    //             {
+                    //                 //update quantity in product database
+                    //                 product_jsonArray.getJSONObject(i).put("quantity", String.valueOf(updated_quantity));
```
Comment (marceloponceardon) [inline src/OrderService/OrderService.java:174]: don't leave giant comments like this

File: src/ProductService/ProductDatabase.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        if (price != 0) {
+            sqlUpdate.append("price = ?, ");
+            valueCount++;
+        }
+        if (quantity != 0) {
+            sqlUpdate.append("quantity = ?, ");
+            valueCount++;
+        }
+
+        // Return 200 to imply that no fields have been updated but still a success (although no change in the db)
```
Comment (marceloponceardon) [inline src/ProductService/ProductDatabase.java:162]: nice

File: tests/combinedWorkload.txt
Comment (marceloponceardon) [inline tests/combinedWorkload.txt:1]: why are you commiting tests/ dir you didn't work on that!!!!
stop using git add all!!!

**Review by marsponce — CHANGES_REQUESTED** [review]
see comments on files

**Review by marsponce — APPROVED** [review]
Looks fine to me, just remember about using tabs to indent 👀

**Pull request merged**